modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentPerVersionStore.java

```diff
@@ -322,13 +322,33 @@ else if (log.isDebugEnabled())
         while (true) {
             List<SharedDeployment> depsToCheck = null;
 
-            SharedDeployment dep = null;
+            SharedDeployment dep;
 
             synchronized (mux) {
                 // Check obsolete request.
                 if (isDeadClassLoader(meta))
                     return null;
 
+                Collection<GridDeployment> created = getDeployments();
+
+                boolean skipSearchDeployment = false;
+
+                // Check already exist deployment.
+                if (meta.deploymentMode() == SHARED) {
+                    for (GridDeployment dep0 : created) {
+                        // hot redeploy from same node
+                        if (dep0.participants().containsKey(meta.senderNodeId()) || dep0.undeployed())
+                            continue;
+
+                        IgniteBiTuple<Class<?>, Throwable> cls = dep0.deployedClass(meta.className(), meta.alias());
+
+                        if (cls.getKey() != null && cls.getValue() == null) {
+                            addParticipant((SharedDeployment)dep0, meta);
```

Comment:
Here you move all partisipants for a class meta to the deploymnt, but bellow we might update the partisipant list in the class meta.
May we try to update the partisipant list at first or don't do the update because the new list won't use?